gh-108191: Add support of positional argument in SimpleNamespace constructor (#108195)
python/cpython: Doc/library/types.rst

   namespace, as well as a meaningful repr.

>    Unlike :class:`object`, with ``SimpleNamespace`` you can add and remove

Reviewer: ```suggestion
   Unlike :class:`object`, with :class:`!SimpleNamespace` you can add and remove
```

   attributes.
   The :class:`!SimpleNamespace` constructor can take a positional argument
   which must be a mapping object with string keys or an :term:`iterable`
   object producing key-value pairs with string keys.
   Key-value pairs from the mapping object or the iterable object are
   directly added to the underlying namespace.
   If a ``SimpleNamespace`` object is initialized with keyword

>    arguments, those are directly added to the underlying namespace.

Reviewer: Should the kwarg behaviour come first, as the longer-established case? I also think initialisation/arguments should be split to a new paragraph, for readability.

Author: I would be grateful if you simply write new documentation. I understand that my writing is not good, but it is the best I can do.

Reviewer: I've suggested a re-wording!

>    The :class:`!SimpleNamespace` constructor can take a positional argument
>    which must be a mapping object with string keys or an :term:`iterable`
>    object producing key-value pairs with string keys.
>    Key-value pairs from the mapping object or the iterable object are
>    directly added to the underlying namespace.
>    If a ``SimpleNamespace`` object is initialized with keyword
>    arguments, those are directly added to the underlying namespace.

Reviewer: Perhaps:
```suggestion

   :py:class:`SimpleNamespace` objects may be initialized either
   with keyword arguments or with a single positional argument.
   When initialized with keyword arguments,
   those are directly added to the underlying namespace.
   Alternatively, when initialized with a positional argument,
   the underlying namespace will be updated with key-value pairs
   from that argument (either a mapping object or
   an :term:`iterable` object producing key-value pairs).
   All such keys must be strings.
```

Author: Thank you!

   A simple :class:`object` subclass that provides attribute access to its
   namespace, as well as a meaningful repr.

   Unlike :class:`object`, with :class:`!SimpleNamespace` you can add and remove
   attributes.

   :py:class:`SimpleNamespace` objects may be initialized
   in the same way as :class:`dict`: either with keyword arguments,
   with a single positional argument, or with both.
   When initialized with keyword arguments,
   those are directly added to the underlying namespace.
   Alternatively, when initialized with a positional argument,
   the underlying namespace will be updated with key-value pairs
   from that argument (either a mapping object or
   an :term:`iterable` object producing key-value pairs).
   All such keys must be strings.

   The type is roughly equivalent to the following code::

       class SimpleNamespace:
           def __init__(self, mapping_or_iterable=(), /, **kwargs):
               self.__dict__.update(mapping_or_iterable)
               self.__dict__.update(kwargs)

>            def __repr__(self):

Reviewer 2: It may be outside the scope of this PR, but I noticed that in gh-109175 you didn't add `__replace__()` to the "equivalent" code here.  It's probably fine that this isn't an exact equivalence, especially since we say "roughly equivalent".

I just wanted to mention it (before I forget), in case it's something that would make sense to address in this PR.  I expect it would basically be what you wrote in https://github.com/python/cpython/issues/109174#issuecomment-1712430549.

Either way, I'll leave it up to you.

Author: It also does not have `__reduce__()`. Both methods would be simple, but I am not sure that it is needed to add them in this example. They may shadow the main idea of the class.

   A simple :class:`object` subclass that provides attribute access to its
   namespace, as well as a meaningful repr.

   Unlike :class:`object`, with :class:`!SimpleNamespace` you can add and remove
   attributes.

   :py:class:`SimpleNamespace` objects may be initialized either

>    with keyword arguments or with a single positional argument.

Reviewer 2: ```suggestion
   :py:class:`SimpleNamespace` objects may be initialized either
   with keyword arguments, with a single positional argument, or with both.
```

Reviewer 2: Perhaps even explicitly point out the similarity to the `dict()` signature, since that may help make thing more clear for readers.

```suggestion
   :py:class:`SimpleNamespace` objects may be initialized
   in the same way as :class:`dict`: either with keyword arguments,
   with a single positional argument, or with both.
```

      Attribute order in the repr changed from alphabetical to insertion (like
      ``dict``).

   .. versionchanged:: 3.13

>       Added support of an optional positional argument.

Reviewer: ```suggestion
      Added support for an optional positional argument.
```

or 

```suggestion
      Support an optional positional argument.
```
